0004-post-tags.md

```diff
@@ -0,0 +1,299 @@
+
+# Rationale and alternatives
+
+The proposed system aims to be a balance between the extremely limited flair feature of reddit and the often fragmented tag system of image hosting platforms, combining the best properties of both options into a theoretically ideal version of the system. Additionally the current system with a singular NSFW flag is far too inflexible to properly moderate the scope of content that falls under NSFW. Many instances have implemented a blanket NSFW ban because of this, greatly limiting what content can be posted where. Aside from that the inability to better sort content is a often discussed topic, especially story communities are desperately craving for a better option to both sort posts based on story association as well as hide/blur posts if they contain spoilers. Given the very high community demand not implementing a tag system of some sort will likely lead to a (soft-)fork of lemmy. The feature is extremely desired by many communities. An implementation of tags via UI rendering ([] brackets for tags in text would turn into a tag on a post) would be possible but ultimately far more limiting and would lead to an incosistent experience across lemmy apps/frontends as well as an inability to properly federate the tags.
+
+# Prior art
+
+Similar implementations can be found on Reddit as well as a great many image hosting platforms.
+
+Reddit's Flair System is rather restrictive, a Post can only have a single custom Flair (such as "News") greatly limiting the usefullness of the feature.
+
+Pros:
+- low moderation effort due to predefined tags
+- no/little content fragmentation since typos are not possible
+- no flair spam due to only 1 flair being allowed at a time
+
+Cons:
+- extreme moderation effort for 1st time setup due to inability to mix flairs
+- filter lists get extremely long because all variants of a flair need to be filtered (for example "News Europe", "News America", etc. instead of just "News")
+- flairs need to be created before being usable
+
+Tag/Flair Systems on Image hosting platform generally tend to favor a more unmoderated experience, often allowing Users to type in the tags themselves.
+
+Pros:
+- enhanced user freedom when applying tags
+- no delay between arising demand for tag and creation of the tag
+
+Cons:
+- greater moderation demand since tags have to be checked for inappropriate content
+- very high chance for content fragmentation (typos, differences in naming things, language barriers)
+- almost completely removes the ability of users to blacklist content tags (pronography would slip past a "pornography" blacklist).
+
+# Unresolved questions
+
+- How would tags federate/display on other Fediverse services such as Mastodon?
```

Comment:
While it'd definitely be useful to federate tags as hashtags and such, I don't think Lemmy should restrict itself to the limitations of hashtags. Instead, consider creating an AP extension (or adopting an already existing extension if one already exists) that:

- Allows editing of the tags themselves (renaming or otherwise)
- Has the ability to adjust tags across instance boundaries (potentially via `Accept/` `Reject/` so implementations can create some sort of "tag suggestion" UI?)
- That does not have anything to do with communities (remember: not all fediverse software need that concept!)
  - The concept of "instance-level tags" and "community-level tags" should be a concept enforced by the Lemmy backend and not ActivityPub itself

This would make it so that any other fediverse projects that want a similar level of flexibility in tags can just adopt Lemmy's extension and get interoperability "for free", even if they're built for different audiences.

Perhaps some tags could be marked as "hashtaggable", but I don't think all tags should federate as hashtags. a tag such as `Spoiler` does not need a hashtag. (in fact, a CW would be more appropriate!), not to mention the more hashtags you add the more Update activities you need to broadcast per tag change.

Reply:
It may be worth writing an [FEP](https://codeberg.org/fediverse/fep/) about this.

Reply:
I saw several discussions around content labeling, and [Web Annotation Protocol](https://www.w3.org/TR/annotation-protocol/) is often brought up. It might be a good foundation for such FEP. No need to adopt the protocol as is, but I think their vocabulary can be useful.